### PR TITLE
Revamped Tutorial

### PR DESCRIPTION
**Pull request by MalkistCoder**
I had a hard time initially when starting with Mineflayer reading through the tutorial on the website, so I tried to make the tutorial more beginner friendly, so that other people starting their journey with Mineflayer get a quick start into learning how to use it.

- Removed Learn JavaScript as it took up most of the document, learning JavaScript can be outsourced to other places.
- Added section that links to courses/books about JavaScript.
- Revamped sections slightly, hopefully making it more beginner friendly

I may add a few more examples later when I have more freetime.

Please give feedback, such as me adding or removing a section, changing a few words, etc. I want to help make the Mineflayer tutorial less daunting and more beginner-friendly to people who already know a bit about JavaScript.

### CODE REVIEW

File: docs/tutorial.md [hunk 1]
```diff
@@ -2,593 +2,253 @@
...
-const options = {
-  host: 'localhost', // Change this to the ip you want.
-  port: 25565 // Change this to the port you want.
-}
-
-const bot = mineflayer.createBot(options)
+const mineflayerPathfinder = require('mineflayer-pathfinder')
+/* Note: It is good practice to keep all your modules
+         the top of your file. Try to put this under
+         the line you imported Mineflayer.
```
Comment (extremeheat) [inline docs/tutorial.md:135]: Try something like "Make sure to place all your require() imports at the top of the file where possible. This prevents your code from being blocked loading modules at runtime by loading all the dependencies before your program code is ran."

to better explain what's happening

File: docs/tutorial.md [hunk 2]
```diff
@@ -2,593 +2,253 @@
...
-So far we have only worked with numbers, but Javascript can work with more variable types:
+This will put Mineflayer in the modules we have installed, and we can start using Mineflayer!
 
-- A string is a piece of text that can contain multiple characters. Strings are defined by using the quotes `''`
+### Create Your First Bot
 
-```js
-const string = 'This is a string' // string type
-```
+I would recommend starting a Minecraft server to test your bot on. A good place to host a small, simple server to host your bot on is [Aternos](https://aternos.org), although you could also host your own Minecraft server locally on `localhost` for more configuration, better ping, and access to more plugins if you're using a software like Spigot or PaperMC.
```
Comment (extremeheat) [inline docs/tutorial.md:53]: I would recommend the tutorial focus on a locally hosted server for simplicity purposes. But you can still mention aternos as a good place to host a server.

**Review by extremeheat — CHANGES_REQUESTED** [review]
Left some comments, in general I would try to make sure the tutorial is as close to the readme/doc/examples as much as possible, that way it's more likely to be kept up to date.

Also, for images I'd recommend uploading them to the repo to avoid them dying.

File: docs/tutorial.md [hunk 3]
```diff
@@ -2,593 +2,253 @@
 
 **Table of Contents**
 
-- [Basics](#basics)
```
Comment (rom1504) [inline docs/tutorial.md:5]: Instead of removing the js part, can you move it to another file ?

**Comment by rom1504** [comment]
Looks like you mostly just wrote another version of a tutorial. What about putting that in a new file ?

File: docs/tutorial.md (hunk 3 above)
Comment (MalkistCoder) [reply docs/tutorial.md:5]: I feel like its better to make others learn on their own instead of putting it in the document. It's how to use Mineflayer, not how to use JavaScript.

File: docs/tutorial.md [hunk 4]
```diff
@@ -2,593 +2,253 @@
...
-[`bot.chatAddPattern()`](http://prismarinejs.github.io/mineflayer/#/api?id=botchataddpatternpattern-chattype-description) method takes three arguments :
+const mineflayer = require('mineflayer')
+const mineflayerPathfinder = require('mineflayer-pathfinder')
 
-- `pattern` - regular expression (regex) to match chat
-- `chatType` - the event the bot emits when the pattern matches. e.g. "chat" or "whisper"
-- `description` - Optional, describes what the pattern is for
+const pathfinder = mineflayerPathfinder.pathfinder
+const Movements = mineflayerPathfinder.Movements
+const GoalNear = mineflayerPathfinder.goals.GoalNear
```
[the remark this answers is not in the document]
Comment (MalkistCoder) [reply docs/tutorial.md:246]: I replaced both parts with
```js
const { pathfinder, Movements } = mineflayerPathfinder
const { GoalNear } = mineflayerPathfinder.goals
```

File: docs/tutorial.md [hunk 5]
```diff
@@ -2,593 +2,253 @@
...
 
-You can add [Groups and Range](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Groups_and_Ranges) into the `pattern`, then the listener will spread the captured groups into arguments of your callback sequentially.
+const bot = mineflayer.createBot({
+  username: 'MineflayerBot',
+  password: 'abc123',
+  // password is your Microsoft password.
+  // If you are using the bot on a cracked/offline server, you don't need the password field.
+  host: 'localhost',
+  port: 25565, // Replace with the port for your aternos server.
+  version: '1.19.4'
```
[the remark this answers is not in the document]
Comment (MalkistCoder) [reply docs/tutorial.md:255]: Yeah, I was using a 1.19.4 server to test because I was too lazy to update my Paper :P
I removed the version number

File: docs/tutorial.md (hunk 2 above)
Comment (MalkistCoder) [reply docs/tutorial.md:54]: I replaced the section with

I would recommend starting a Minecraft server to test your bot on. A good place to start is to host the server [on your own computer](https://help.minecraft.net/hc/en-us/articles/360058525452-How-to-Setup-a-Minecraft-Java-Edition-Server). If you need to test your bot with plugins, you can use a software such as [PaperMC](https://papermc.io/). If you are too lazy to go through the hassle of hosting locally, you could use a free host such as [Aternos](https://aternos.org/).

**Review by MalkistCoder — COMMENTED** [review]
I have changed the changes in the new version of the tutorial.

File: docs/tutorial.md (hunk 3 above)
Comment (rom1504) [reply docs/tutorial.md:5]: I don't think that makes sense. Most users are blocked by their absence of knowledge of JavaScript

Comment (frej4189) [reply docs/tutorial.md:5]: I have to agree with @MalkistCoder. Documentation should be exclusively relating to mineflayer. Full-fledged tutorials should be extracted to things like blog posts, or, if you really care for it, to a separate repository.

File: docs/tutorial.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (extremeheat) [reply docs/tutorial.md]: Still not resolved, you should copy over what's in the README

File: docs/tutorial.md (hunk 3 above)
Comment (rom1504) [reply docs/tutorial.md:5]: Tutorials are here to help concrete users solve their issues. They are not made for theoretical users.
Most of our users don't know javascript.

**Comment by rom1504** [comment]
Hello,
I made several comments here but I see they are not addressed.
Could you please take another look?

**Comment by MalkistCoder** [comment]
> Hello, I made several comments here but I see they are not addressed. Could you please take another look?

I'll check it out.
Also do you know how to fix this? I already did everything he recommended a while ago but this is still there
https://i.imgur.com/620Pj0k.png

**Review by MalkistCoder — COMMENTED** [review]
how do i finish reviewing?

File: docs/tutorial.md (hunk 3 above)
Comment (MalkistCoder) [reply docs/tutorial.md:5]: Most of the people that use mineflayer know how to code, so I feel like the JavaScript section is not really necessary.

If someone needs/wants to learn JavaScript, I added multiple resources such as JavaScript.info (which uses multiple languages) and Codecademy's JavaScript tutorial (interactive), so even if they don't know JavaScript they can learn it easily and quickly.

Comment (rom1504) [reply docs/tutorial.md:5]: Feel free to read room on discord to convince yourself otherwise
You may also check the issues section in this repository

**Comment by rom1504** [comment]
> I have changed the changes in the new version of the tutorial.

It seems it's still the same file. Create a new file instead

**Comment by rom1504** [comment]
hey,

your tutorial is providing a different introduction to the same thing the old tutorial. Information between them are complementary.
what about keeping the old tutorial exactly as it was and providing your tutorial in a new file ?

**Comment by MalkistCoder** [comment]
> hey,
> 
> your tutorial is providing a different introduction to the same thing the old tutorial. Information between them are complementary. what about keeping the old tutorial exactly as it was and providing your tutorial in a new file ?

thats why i made a new tutorial, the introduction is too long, its better to make other users explore the way they learn JavaScript by going to other sites. imagine if every single package had to explain what every single javascript feature is before you get to the tutorial/api reference. i had such a hard time getting into mineflayer using the tutorial page because there was so much random javascript tutorial at the start. i think giving the job to other websites/teaching sites who are more dedicated to teaching javascript is a better idea

basically, i think people come to the tutorial to learn mineflayer, not to learn javascript. i have already provided the old tutorial in the tutorial.md file if people prefer it more than going to another site to learn it. i can see you want people that dont know javascript to learn it through the tutorial, but there are already many dedicated websites to do the task, so placing a massive javascript tutorial that is larger than the actual mineflayer section isnt necessary.

**Comment by rom1504** [comment]
do you see any drawback in keeping both your tutorial and the old one ? that way users can choose the one they prefer

**Comment by MalkistCoder** [comment]
> do you see any drawback in keeping both your tutorial and the old one ? that way users can choose the one they prefer

I put the old JavaScript tutorial in a different file (jstutorial.md). Other than that I made the tutorial a more beginner-friendly version than the old tutorial, so I don't feel the need to include the full old mineflayer tutorial.

**Comment by rom1504** [comment]
The new tutorial is missing a lot of information that the old one contains. For example see the section about promises.

Here are he options:
* you include all the infos the old tutorial had
* you keep the old tutorial and add yours in a new file
* you close the PR

**Comment by MalkistCoder** [comment]
Ill add the promises section to `jstutorial.md`, let me know if theres any more things from the old tutorial you would like me to add

File: docs/tutorial.md [hunk 6]
```diff
@@ -2,621 +2,388 @@
...
+Then, we will use the TypeScript compiler to create a `tsconfig.json` file.
 
-- `bot.on(eventName, listener)`
-  Execute the `listener` function for each time the event named `eventName` triggered.
-- `bot.once(eventName, listener)`
-  Execute the `listener` function, only once, the first time the event named `eventName` triggered.
-- `bot.removeListener(eventName, listener)`
-  Removes the specified `listener` for the event named `eventName`. In order to use this you either need to define your function with `function myNamedFunc() {}` or put your function in a variable with `const myNamedFunc = () => {}`. You can then use `myNamedFunc` in the listener argument.
-
-Not only bot object, [`Chest`](http://prismarinejs.github.io/mineflayer/#/api?id=mineflayerchest), [`Furnace`](http://prismarinejs.github.io/mineflayer/#/api?id=mineflayerfurnace), [`Dispenser`](http://prismarinejs.github.io/mineflayer/#/api?id=mineflayerdispenser), [`EnchantmentTable`](http://prismarinejs.github.io/mineflayer/#/api?id=mineflayerenchantmenttable), [`Villager`](http://prismarinejs.github.io/mineflayer/#/api?id=mineflayervillager) object also have their own events!
```
Comment (rom1504) [inline docs/tutorial.md:431]: this is not present in the new tutorial

File: docs/tutorial.md [hunk 7]
```diff
@@ -2,621 +2,388 @@
...
+```bash
+npx tsc --init
+```
 
-### Promises
-A [promise](https://nodejs.dev/learn/understanding-javascript-promises) is a function that you can use the `await` variable to wait on until it's job is complete. (you can omit the await to not wait for results)
```
Comment (rom1504) [inline docs/tutorial.md:434]: this is not present in the new tutorial

File: docs/tutorial.md [hunk 8]
```diff
@@ -2,621 +2,388 @@
...
 }
 ```
 
-The above code will try to consume what the bot is currently holding.  
-When the consuming ends, the function that is passed along is called.  
-We can then do other things that we want to do after.  
-The function could also be called when an error occurs.
+Now, rename your `index.js` file to `index.ts` to change it to TypeScript. Do this with all of your code. Then, go in your command line and type:
 
-#### Correct and incorrect approach
```
Comment (rom1504) [inline docs/tutorial.md:452]: this is not present in the new tutorial

File: docs/tutorial.md [hunk 9]
```diff
@@ -2,621 +2,388 @@
...
-| Value | `node` | `filename.js` | `host` | `port` | `username` | `password` |
-| Index | `[0]` | `[1]` | `[2]` | `[3]` | `[4]` | `[5]`
+[Full code](#find-me-bot)
 
-### Passing along functions
+![The bot found me!](img/foundme.png)
 
-Not only basics variables like numbers and strings can be given as an argument.  
-Functions can also be passed as a variable.
+## Adding TypeScript
```
Comment (rom1504) [inline docs/tutorial.md:240]: I don't think we should recommend typescript

File: docs/tutorial.md [hunk 10]
```diff
@@ -2,621 +2,388 @@
...
 ```
 
-This will create a new variable named `test` and assign it the value `5`  
-Variable are used to save data and use it later in the code.
-
-Now save the file so we can run the code. Open a terminal again (or a new terminal in VSCode) and navigate to the same folder the file is saved in. This can be done using the `cd` command, for example: `cd Documents\javascript`  
-Once your terminal is in the same folder as your Javascript file, you can run `node filename.js`  
-If you have done everything correctly, you should see nothing.  
-In the next chapter we will show you how you can 'print' things to the terminal.
+This will create a `project.json` file that will store all our project's packages, so that if you want to send your code to someone else, they can easily download Mineflayer and any other dependencies you need.
```
Comment (rom1504) [inline docs/tutorial.md:48]: no package.json

File: docs/tutorial.md [hunk 11]
```diff
@@ -2,621 +2,388 @@
...
-Javascript is then able to make your code run more efficiently because it knows it doesn't have to account for value changes for that variable.  
-If you want a modifiable variable, you will still have to use `let` of course.
+Secondly, we need to install the `mineflayer` package from the Node Package Manager (NPM). :
 
-```js
-const test = 5
-// eslint-disable-next-line
-test = 10 // This line is invalid.
+```bash
+npm i mineflayer
```
Comment (rom1504) [inline docs/tutorial.md:53]: should be npm i --save mineflayer

**Comment by rom1504** [comment]
added a few comments

Overall it's not clear to me who this tutorial is targeting

**Comment by MalkistCoder** [comment]
I probably wont be able to keep contributing this due to events right now in my personal life, but I hope someone else can make the tutorial more simple and concise.
